src/stable_vec.rs

```diff
@@ -0,0 +1,396 @@
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+/// A key used to access an element in the StableVec. Currently, this is just a `usize` index into a
+/// backing array, but this is an implementation detail and may eventually change. In particular
+/// this may become an opaque type which holds a generational index for bug-catching purposes.
+pub type Key = usize;
+const INVALID_KEY: Key = Key::MAX;
+
+#[derive(Clone)]
+enum Entry<T> {
+    /// An occupied entry.
+    Used {
+        /// The value contained by the entry.
+        contents: T,
+        // TODO: Should I implement a generational index for bug-catching? If added, I should make
+        // the raw index u32 and the generational index u32 so they can fit in a single 64 bit
+        // register, and specifically stick the generational index in the high 32 bits so it can be
+        // easily masked out for indexing. The generational index should live outside of the enum.
+    },
+    /// An unoccupied entry.
+    Free {
+        /// The index of the next free entry in the data structure's internal storage. This induces
+        /// a chain of free entries (a "free list"). The chain terminates when the contained value
+        /// is [`INVALID_KEY`].
+        next: Key,
```

Comment:
This almost certainly doesn't matter, but it occurs to me that this could be a good use case for `Option<Key>` if `Key` were actually a [NonZeroUsize](https://doc.rust-lang.org/stable/std/num/struct.NonZeroUsize.html). Notably, `Option<Key>` would then be the same size as `Key`, taking care of the in-band signaling in a way that shows up in the type system.

Reply:
I considered that, but thought it would require filling entry 0 with a default value, which would restrict the collection to `impl Default` types. But obviously that's not the case, I can just add a free entry there and never reference it! Not sure what I was thinking...

Reply:
That's a very good point!